napari/_qt/_qapp_model/injection/_qproviders.py

```diff
@@ -0,0 +1,34 @@
+"""Qt providers.
+
+Non-Qt providers can be found in `napari/_app_model/injection/_providers.py`.
+"""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Optional
+
+if TYPE_CHECKING:
+    from napari._qt.qt_main_window import Window
+    from napari._qt.qt_viewer import QtViewer
+
+
+def _provide_qt_viewer() -> Optional[QtViewer]:
+    from napari._qt.qt_main_window import _QtMainWindow
+
+    if _qmainwin := _QtMainWindow.current():
+        return _qmainwin._qt_viewer
+    return None
+
+
+def _provide_window() -> Optional[Window]:
+    from napari._qt.qt_main_window import _QtMainWindow
+
+    if _qmainwin := _QtMainWindow.current():
+        return _qmainwin._window
+    return None
```

Comment:
So there can be more than one provider per type hint and `in-n-out` will iterate through providers. Providers are meant to return `None` if it is not able to return the object so it can continue iterating. `in-n-out` itself will raise an error if it is not able to provide something.

I recently amended their docs to try to add this info, if you want more details: https://ino.readthedocs.io/en/latest/getting_started/#injecting-dependencies-into-functions